src/txmempool.cpp

```diff
@@ -2009,6 +2063,80 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
           cachedInnerUsage);
 }
 
+bool CTxMemPool::trimToSize(const CMemPoolEntry* entry, size_t max_size) {
+    LOCK(cs);
+    size_t new_entry_usage = entry ? entry->GetSize() : 0;
+    assert(new_entry_usage < max_size);
+    size_t current_usage = totalTxSize + totalCertificateSize;
+    if (current_usage + new_entry_usage <= max_size) return true;
+
+    const CCertificateMemPoolEntry* certificate = dynamic_cast<const CCertificateMemPoolEntry*>(entry);
+    if (certificate && totalTxSize > m_max_size / 2) {
+        LogPrint("mempool", "%s():%d - Trying to remove transactions to make room for certificate (tot_tx: %zu, max: %zu)\n", __func__, __LINE__, totalTxSize, m_max_size / 2);
+        trimToSize(nullptr, current_usage - new_entry_usage);
+        current_usage = totalTxSize + totalCertificateSize;
+        if (current_usage + new_entry_usage <= m_max_size) return true;
+    }
+
+    // we must either reject incoming tx, or remove something
+    int64_t size_to_be_removed = current_usage + new_entry_usage - max_size;
+    std::multimap<CRawFeeRate, uint256> mapTxRates;
+    std::map<uint256, CRawFeeRate> avgRates;
+    if (certificate) {
+        LogPrint("mempool", "%s():%d - Trying to remove other certificates to make room for certificate\n", __func__, __LINE__);
+        for (const auto& cert: mapCertificate) {
+            CRawFeeRate feerate = avgFeeRateWithDeps(cert.second.GetCertificate(), true, avgRates);
+            mapTxRates.insert({feerate, cert.first});
+        }
+    }
+    else {
+        LogPrint("mempool", "%s():%d - Trying to remove transactions to make room\n", __func__, __LINE__);
+        for (const auto& tx: mapTx) {
+            CRawFeeRate feerate = avgFeeRateWithDeps(tx.second.GetTx(), true, avgRates);
+            mapTxRates.insert({feerate, tx.first});
+        }
+    }
+
+    // Check what should be removed, and if this selection includes entry
+    // If it includes entry, then just reject the incoming transaction and do nothing else
+    std::vector<const CMemPoolEntry*> to_be_removed;
+    for (auto remove_candidate = mapTxRates.begin(); remove_candidate != mapTxRates.end() && size_to_be_removed > 0; ++remove_candidate) {
+        if (entry && remove_candidate->first >= CFeeRate(entry->GetFee(), entry->GetSize())) return false;
+
+        const CMemPoolEntry* t = certificate ? static_cast<CMemPoolEntry*>(&mapCertificate.at(remove_candidate->second)) : static_cast<CMemPoolEntry*>(&mapTx.at(remove_candidate->second));
+        to_be_removed.push_back(t);
+        size_to_be_removed -= remove_candidate->first.GetBytes();
+    }
+
+    // Actually remove things from mempool
+    unsigned int total_removed = 0;
+    for (const CMemPoolEntry* r: to_be_removed) {
+        std::list<CTransaction> removed_txs;
+        std::list<CScCertificate> removed_certs;
+        if (certificate) {
```

Comment:
This condition is wrong, this check should be on `r`. Will fix this with the next push, sorry about that.